cirq-google/cirq_google/workflow/quantum_executable.py

```diff
@@ -0,0 +1,157 @@
+            params: A cirq.ParamResolverOrSimilarType which will be frozen into a tuple of
+                key value pairs.
+            spec: Specification metadata about this executable that is not used by the quantum
+                runtime, but is persisted in result objects to associate executables with results.
+            problem_topology: Description of the multiqubit gate topology present in the circuit.
+                If not specified, the circuit must be compatible with the device topology.
+            initial_state: How to initialize the quantum system before running `circuit`. If not
+                specified, the device will be initialized into the all-zeros state.
+        """
+
+        # We care a lot about mutability in this class. No object is truly immutable in Python,
+        # but we can get pretty close by following the example of dataclass(frozen=True), which
+        # deletes this class's __setattr__ magic method. To set values ever, we use
+        # object.__setattr__ in this __init__ function.
+        #
+        # We write our own __init__ function to be able to accept a wider range of input formats
+        # that can be easily converted to our native, immutable format.
+        object.__setattr__(self, 'circuit', circuit.freeze())
+
+        if not isinstance(measurement, BitstringsMeasurement):
+            raise ValueError(f"`measurement` should be a BitstringsMeasurement, not {measurement}.")
+        object.__setattr__(self, 'measurement', measurement)
+
+        if isinstance(params, tuple) and all(
+            isinstance(param_kv, tuple) and len(param_kv) == 2 for param_kv in params
+        ):
+            frozen_params = params
+        elif isinstance(params, Sequence) and all(
+            isinstance(param_kv, Sequence) and len(param_kv) == 2 for param_kv in params
+        ):
+            frozen_params = tuple((k, v) for k, v in params)
+        elif _is_param_resolver_or_similar_type(params):
+            param_resolver = cirq.ParamResolver(cast(cirq.ParamResolverOrSimilarType, params))
+            frozen_params = tuple(param_resolver.param_dict.items())
+        else:
+            raise ValueError(f"`params` should be a ParamResolverOrSimilarType, not {params}.")
+        object.__setattr__(self, 'params', frozen_params)
+
+        if not isinstance(spec, ExecutableSpec):
+            raise ValueError(f"`spec` should be an ExecutableSpec, not {spec}.")
+        object.__setattr__(self, 'spec', spec)
+
+        if problem_topology is not None and not isinstance(problem_topology, NamedTopology):
+            raise ValueError(
+                f"`problem_topology` should be a NamedTopology, " f"not {problem_topology}."
+            )
+        object.__setattr__(self, 'problem_topology', problem_topology)
+
+        if initial_state is not None and not isinstance(initial_state, cirq.ProductState):
+            raise ValueError(f"`initial_state` should be a ProductState, not {initial_state}.")
```

Comment:
Nit: I think a lot of these error catching mechanisms here need to be reworked to avoid making violation checking of the API part of the API. *"You should not document exceptions that get raised if the API specified in the docstring is violated (because this would paradoxically make behavior under violation of the API part of the API)"* https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

Reply:
I want to provide as strong a guarantee as possible in python that this is immutable. It's only immutable if its members are immutable. Do you have any better ideas for validating that input objects are immutable?

Reply:
I'm not sure I follow... I was more pointing out that we don't need so many error checks in `__init__`. From the code, If you take:
```python3
if not isinstance(measurement, BitstringsMeasurement):
    raise ValueError(f"`measurement` should be a BitstringsMeasurement, not {measurement}.")
object.__setattr__(self, 'measurement', measurement)
```
it should just turn into:
```python3
object.__setattr__(self, 'measurement', measurement)
```

Reply:
Then someone could pass in anything as `measurement` including a custom, mutable class with a `n_repetitions` attribute and now my QuantumExecutable is subtly mutable.

These checks are for input validation. The style guide you quote (that I have brought up many times) is saying not to **document input validation**. It is generally good style to **do input validation**.

Reply:
> It is generally good style to do input validation.

I don't disagree.

> Then someone could pass in anything as measurement including a custom, mutable class with a n_repetitions attribute and now my QuantumExecutable is subtly mutable.


So we want to prohibit some user from matching the `BitStringMeasurement` interfaces and discourage the idea of duck typing here in the pursuit of immutability ?

Reply:
For this specific check, my prototype has more measurement types, so I wanted to check it was one of the finite options. I'll remove now and maybe re-introduce when there are more options. 

This constructor will throw if any of the fields don't support `hash()` which is as good as it's gonna get for duck typing immutability in Python. Related shout out to #4327 

I've also removed the `spec`, `problem_topology`, `initial_state` isinstance check, since duck typing would suffice.

The remaining arguments have "freezing" logic.